Add the Signer interface

MVrachev opened the pull request:
Fixes: https://github.com/theupdateframework/tuf/issues/1263?notification_referrer_id=MDE4Ok5vdGlmaWNhdGlvblRocmVhZDE0ODExNjA4Mjg6MTYyNDY3Nzg%3D#issuecomment-772616454

### Description of the changes being introduced by the pull request:

Create a Signer interface with the purpose of supporting multiple signing implementations.
Read more on the issue above.

### Please verify and check that the pull request fulfils the following requirements:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

trishankatdatadog commented on tests/test_signer.py:
```diff
@@ -0,0 +1,96 @@
+  def test_sslib_sign(self):
+    # Test generation of RSA signatures.
+    sslib_signer = SSlibSigner(self.rsakey_dict)
+    rsa_sig_obj = sslib_signer.sign(self.DATA)
```
I like how simple it looks to use, judging by the code so far! 👏🏽

trishankatdatadog commented on securesystemslib/signer.py:
```diff
@@ -0,0 +1,131 @@
+      metadata related to the key needed for verification (e.g. keyid).
+    """
+    signiture_dict = sslib_keys.create_signature(self.key_dict, payload)
+    return Signature(signiture_dict['keyid'], signiture_dict['sig'])
```
Hmm, why this indirection? To keep backwards-compatibility for existing functions?

MVrachev replied:
What do you mean by `indirection`? 

PS: I will push a new version of this with the `**` syntax. It looks a lot cleaner.

MVrachev commented:
This interface is ready for review @lukpueh!
I will add another pr for GPGSigner.

lukpueh commented on tests/test_signer.py:
```diff
@@ -0,0 +1,102 @@
+
+    # Supplying an incorrect number of arguments.
+    self.assertRaises(TypeError, sslib_signer.sign)
+    self.ecdsakey_dict['keyval']['private'] = private
```
You could DRY the test e.g. with a for loop over `[self.rsakey_dict, self.ecdsakey_dict]`. Then you could easily throw the currently unused `self.ed25519key_dict` in the mix without extra effort. :)

lukpueh commented on securesystemslib/signer.py:
```diff
@@ -0,0 +1,127 @@
+    self.signature = sig
+
+
+  def to_dict(self):
```
If you provide a `to_dict`, I suggest you also provide a `from_dict` classmethod.

lukpueh reviewed (changes requested):
Nice work, @MVrachev. I only pointed out some minor issues, most of which are not really your fault but seem to stem from inaccurate code documentation in this repo.

lukpueh commented on securesystemslib/signer.py:
```diff
@@ -0,0 +1,127 @@
+      Returns a "Signature" class instance containing the signature and the
+      the keyid which uniquely identifies the key used for signature generation.
+    """
+    pass
```
This should raise `NotImplementedError`

MVrachev commented on tests/test_signer.py:
```diff
@@ -0,0 +1,102 @@
+
+    # Verify rsa signature
+    verified = KEYS.verify_signature(self.rsakey_dict,
+        rsa_sig_obj.to_dict(), self.DATA)
```
A mistake it seems. I will remove it.

MVrachev commented on tests/test_signer.py:
```diff
@@ -0,0 +1,102 @@
+
+    # Verify rsa signature
+    verified = KEYS.verify_signature(self.ecdsakey_dict,
+        ecdsa_sig.to_dict(), self.DATA)
```
A mistake it seems. I will remove it.

MVrachev commented on tests/test_signer.py:
```diff
@@ -0,0 +1,102 @@
+    self.rsakey_dict['keyval']['private'] = ''
+    sslib_signer.key_dict = self.rsakey_dict
+
+    self.assertRaises(ValueError, sslib_signer.sign, self.DATA)
```
A mistake. It should raise a `ValueError` for the missing value.
Will fix it.

MVrachev commented:
Updated the pr.
- I added a new commit fixing the documentation for `secureystemslib.keys.create_signature()`. 
- Added a `Signature.from_dict()` and a unit test for it.
- Fixed the wording in the documentation
- Rewrote the test and made it simpler.

@lukpueh sorry for the obvious mistakes in the tests. I copied a lot from `tests/test_keys.py` and it seems I did it badly. :D

lukpueh commented on securesystemslib/signer.py:
```diff
@@ -0,0 +1,144 @@
+      Creates a Signature object from its JSON/dict representation.
+
+    <Arguments>
+      signature_dict: A dict containing that should contain valid keyid
```
```suggestion
      signature_dict: A dict that contains a valid keyid
```
?

lukpueh commented on securesystemslib/signer.py:
```diff
@@ -0,0 +1,144 @@
+
+    <Returns>
+      Returns a "Signature" class instance containing the signature and the
+      the keyid which uniquely identifies the key used for signature generation.
```
```suggestion
     keyid which uniquely identifies the key used for signature generation.
```

lukpueh commented on securesystemslib/signer.py:
```diff
@@ -0,0 +1,144 @@
+      Securesystemslib default implementation of the "Signer" interface.
+      With this implementation the following signature schemes are supported:
+
+      'RSASSA-PSS'
```
We actually support pkcs1v15 as well. And I think we should use the securesystemslib scheme strings here, as e.g. defined in formats:

```python
 RSA_SCHEME_SCHEMA = SCHEMA.OneOf([
  SCHEMA.RegularExpression(r'rsassa-pss-(md5|sha1|sha224|sha256|sha384|sha512)'),
  SCHEMA.RegularExpression(r'rsa-pkcs1v15-(md5|sha1|sha224|sha256|sha384|sha512)')])
```

lukpueh commented on tests/test_signer.py:
```diff
@@ -0,0 +1,87 @@
+
+
+  def test_sslib_sign(self):
+    for scheme_dict in [self.rsakey_dict, self.ecdsakey_dict]:
```
ping `ed25519key_dict`?

lukpueh reviewed (changes requested):
Thanks, @MVrachev! This is getting better and better. Here are a few more suggestions.

MVrachev commented on securesystemslib/keys.py:
```diff
@@ -713,9 +716,9 @@ def create_signature(key_dict, data):
+  # Continue to support keytypes of ecdsa-sha2-nistp256
   # for backwards compatibility with older securesystemslib releases
-  elif keytype in ['ecdsa', 'ecdsa-sha2-nistp256', 'ecdsa-sha2-nistp384']:
+  elif keytype in ['ecdsa', 'ecdsa-sha2-nistp256',]:
```
If the `keytype` is `ecdsa-sha2-nistp384` the execution will continue with the call of `securesystemslib.ecdsa_keys.create_signature(public, private, data, scheme)`. 
I am wondering what will be the value of the `scheme` on this call?
Because the `scheme` argument is the one used to determine if the `ecdsa-sha2-nistp384` signature generation is actually supported.

If the scheme is again  `ecdsa-sha2-nistp384`, then there is no sense in keeping this comment and check and how to drop support for `ecdsa-sha2-nistp256` and `ecdsa-sha2-nistp384` could be decided independently of this.

MVrachev replied:
I decided to not update the commit now until we decide what to do next.

trishankatdatadog replied:
Nitpick: I recommend using a set instead of a list to capture the values to check...

lukpueh replied:
> If the scheme is again ecdsa-sha2-nistp384, then there is no sense in keeping this comment and check and how to drop support for ecdsa-sha2-nistp256 and ecdsa-sha2-nistp384 could be decided independently of this.

Not sure I follow your reasoning, but I don't think this particular line of code is worth any further consideration without taking into account the big picture in #308 or the even bigger picture in https://github.com/secure-systems-lab/signing-spec/issues/9.

joshuagl replied:
> Long story short, instead of making this unrelated change now, I'd rather think about how to best deprecate both 'ecdsa-sha2-nistp256' and 'ecdsa-sha2-nistp384' as keytypes.

Yes please, let's figure out how to deprecate both 'ecdsa-sha2-nistp256' and 'ecdsa-sha2-nistp384' as keytypes.

> @joshuagl, do you want to weigh in as the author of that last change?

With or without this change we don't support generating signatures with a keytype of ecdsa-sha2-nistp384, so I don't feel too strongly about how it fails. It was a mistake to introduce the change in #228, but let's not worry too much about trying to undo it now and focus on the bigger picture items and deprecating the overloaded keytypes.

MVrachev replied:
Okay, then I will leave this commit for now.

MVrachev commented on securesystemslib/keys.py:
```diff
@@ -684,7 +687,7 @@ def create_signature(key_dict, data):
   # Signing the 'data' object requires a private key. Signing schemes that are
   # currently supported are: 'ed25519', 'ecdsa-sha2-nistp256',
-  # 'ecdsa-sha2-nistp384' and rsa schemes defined in
+  # and rsa schemes defined in
```
I updated it above adding the `ecdsa-sha2-nistp256` support.

MVrachev replied:
Still, I added 
`rsassa-pss-(md5|sha1|sha224|sha256|sha384|sha512)`
`rsa-pkcs1v15-(md5|sha1|sha224|sha256|sha384|sha512)`
in the doc.

MVrachev commented:
I rebased and squashed the commits where I address @lukpueh  comments into the appropriate commits. 
I added the `ed25519key_dict` into the tests and fixed another group of docstring mistakes.

lukpueh reviewed (changes requested):
I think we should use the same strategy to move away from legacy code style as we do in TUF (see [ADR 5](https://github.com/theupdateframework/tuf/blob/develop/docs/adr/0005-use-google-python-style-guide.md)). So given that you are adding completely new modules I suggest you also follow the new [style guideline](https://github.com/secure-systems-lab/code-style-guidelines/blob/master/python.md), especially in regards to code documentation. What do you think?

MVrachev commented:
> I think we should use the same strategy to move away from legacy code style as we do in TUF (see [ADR 5](https://github.com/theupdateframework/tuf/blob/develop/docs/adr/0005-use-google-python-style-guide.md)). So given that you are adding completely new modules I suggest you also follow the new [style guideline](https://github.com/secure-systems-lab/code-style-guidelines/blob/master/python.md), especially in regards to code documentation. What do you think?

I thought I should follow the style guidelines in the other parts in securesytemslib.
My reasoning was that in tuf where we write the new API from scratch we should use the new style guidelines, but for files that are in the old folders, we should use the older style guidelines. 

It seems I am wrong. I will fix it on Monday.

lukpueh commented:
Your reasoning makes perfect sense, @MVrachev, especially if you read ADR 5 together with [ADR 3](https://github.com/theupdateframework/tuf/blob/develop/docs/adr/0003-where-to-develop-TUF-1-0-0.md). However, ADR 5 alone even prescribes the new style for "new files" and not only for "new files in a certain new directory".

But regardless of ADR 5, I think it makes a lot of sense to use the new docstring format everywhere especially for classes / functions that should be public API (I even adopted it in an existing module recently, see #288). 

I am less passionate about most of the other style recommendations, but to make things easier I would just follow the "new style for new files" rule of thumb.

MVrachev commented:
I updated the pr with the new style guidelines.

MVrachev commented:
I had to force push to resolve conflicts.
I updated the `Update doc for ecdsa-sha2-nistp384 schema` commit to using a `keys_types_set` as Trishank suggested.

lukpueh commented on securesystemslib/signer.py:
```diff
@@ -0,0 +1,137 @@
+
+
+class SSlibSigner(Signer):
+    """
+    Purpose:
+        Securesystemslib default implementation of the "Signer" interface.
+        With this implementation the following signature schemes are supported:
+
+        'rsassa-pss-(md5|sha1|sha224|sha256|sha384|sha512)'
+        'rsa-pkcs1v15-(md5|sha1|sha224|sha256|sha384|sha512)'
+        RFC3447 - RSASSA-PSS
+        http://www.ietf.org/rfc/rfc3447.
+
+        'ed25519'
+        ed25519 - high-speed high security signatures
+        http://ed25519.cr.yp.to/
+
+        'ecdsa-sha2-nistp256'
+        https://tools.ietf.org/html/rfc5656
+
+    Attributes:
+        key_dict:
+            A dictionary containing the keys. Both private and public keys are
+            needed.
+            Which signature to generate is determined by the key type of
+            'key_dict'.
+
+            An example RSA key dict has the form:
+
+            {'keytype': 'rsa',
+            'scheme': 'rsassa-pss-sha256',
+            'keyid': 'f30a0870d026980100c0573bd557394f8c1bbd6...',
+            'keyval': {'public': '-----BEGIN RSA PUBLIC KEY----- ...',
+                        'private': '-----BEGIN RSA PRIVATE KEY----- ...'}}
+
+            The public and private keys are strings in PEM format.
+    """
```
I tried to clarify the parts about key types and schemes a bit and ended up re-writing the entire docstring. What do you think?

```suggestion
    """A securesystemslib signer implementation.

    Provides a sign method to generate a cryptographic signature with a
    securesystemslib-style rsa, ed25519 or ecdsa private key on the instance.
    The signature scheme is determined by the key and must be one of:

    - rsa(ssa-pss|pkcs1v15)-(md5|sha1|sha224|sha256|sha384|sha512) (12 schemes)
    - ed25519
    - ecdsa-sha2-nistp256

    See 'securesystemslib.interface' for functions to generate and load keys.

    Attributes:
        key_dict:
            A securesystemslib-style key dictionary, which includes a keyid,
            key type, signature scheme, and the public and private key values,
            e.g.::

                {
                    "keytype": "rsa",
                    "scheme": "rsassa-pss-sha256",
                    "keyid": "f30a0870d026980100c0573bd557394f8c1bbd6...",
                    "keyval": {
                        "public": "-----BEGIN RSA PUBLIC KEY----- ...",
                        "private": "-----BEGIN RSA PRIVATE KEY----- ..."
                    }
                }

    """
```

Note that the code block now also renders nicely with sphinx.

MVrachev replied:
Looks great. I will use your suggestion.

lukpueh commented on securesystemslib/keys.py:
```diff
@@ -706,9 +711,9 @@ def create_signature(key_dict, data):
+  # Continue to support keytypes of ecdsa-sha2-nistp256
   # for backwards compatibility with older securesystemslib releases
-  elif keytype in ['ecdsa', 'ecdsa-sha2-nistp256', 'ecdsa-sha2-nistp384']:
+  elif keytype in key_types_set:
```
I'm not particularly happy about the unrelated changes in this module, at least not as part of this PR.
Even with your fixes the docstring is still inaccurate (e.g. *"cryptography library specified in 'settings'"*) and could be more concise (see my suggestions in the corresponding `SSlibSigner` docstring). And I'd also have a comment or two on the changes in the function body.

Would you mind just removing 67699608ea017c0a2c9c1dca51f03c2b589bbc53 from this PR? You can instead submit a new issue or add a comment e.g. to https://github.com/secure-systems-lab/securesystemslib/issues/271.

MVrachev replied:
Okay, I will remove the commit.
Added a comment about this: https://github.com/secure-systems-lab/securesystemslib/issues/271#issuecomment-782144697

lukpueh reviewed (changes requested):
Hey Martin, this looks great, I only left a few comments regarding docstrings (format), and I'd like to kindly ask you to remove the changes you made in `securesystemslib/keys.py` from this PR. :)

MVrachev commented on securesystemslib/signer.py:
```diff
@@ -0,0 +1,137 @@
+
+class Signature:
+    """Storage class containing information about a signature and the keyid
+        uniquely identifying the key used to generate the signature.
```
I will update the class docstring here, but I am not sure there is something else to add to the `SSlibSigner` docstring, so
the class summary is actually the class docstring.
The rest information related to the functionality of the class is into the class method `sign`.

MVrachev commented:
Updated the pr addressing @lukpueh comments around the docstrings and tried my best to follow the style guidelines. 
I decided to squash the changes into the `Add the Signer interface and the SSlibSigner` commit in order to prepare the pr for merge.

PS: Sorry that I force-pushed so many times. I always forgot to add something.

lukpueh commented on securesystemslib/signer.py:
```diff
@@ -0,0 +1,137 @@
+Purpose:
+    The goal of this module is to provide signing interface supporting multiple
+    signing implementations.
+"""
```
> Same comment applies to the module docstring in test_signer.

ping

MVrachev commented:
Rebased and update the `test_signature.py` file because of https://github.com/secure-systems-lab/securesystemslib/pull/319#discussion_r580063012.

lukpueh commented on tests/test_signer.py:
```diff
@@ -0,0 +1,88 @@
+import securesystemslib.keys as KEYS
+from securesystemslib.exceptions import FormatError, UnsupportedAlgorithmError
+
+# TODO: Remove case handling when fully dropping support for versions >= 3.6
```
Looks like you copy-pasted a mistake I made in tuf's test_api.py :P
```suggestion
# TODO: Remove case handling when fully dropping support for versions < 3.6
```
```suggestion
# TODO: Remove case handling when fully dropping support for versions >= 3.6
```

lukpueh commented on tests/test_signer.py:
```diff
@@ -0,0 +1,88 @@
+            scheme_dict['keyval']['private'] = ''
+            sslib_signer.key_dict = scheme_dict
+
+            with self.assertRaises((ValueError, FormatError)):
```
Above you say:
> Removing private key from 'scheme_dict' - should raise a ValueError

Why do you assert for both `(ValueError, FormatError)`? Is this different for different keys? 🤦

MVrachev replied:
Yes, sadly.
For `rsa` keys   if there is no private key an `ValueError` is raised, 
for `ecdsa` and `ed25519` is a `FormatError`.

I will fix the comment above which says there should be a `ValueError`.

lukpueh replied:
> I was thinking should I create an issue or pr for this, but given that I am researching how to replaces the securesystemslib schemas probably it doesn't make sense.

I agree. I'll still make a note on #191.

lukpueh commented on tests/test_signer.py:
```diff
@@ -0,0 +1,88 @@
+            scheme_dict['scheme'] = 'invalid_scheme'
+            sslib_signer = SSlibSigner(scheme_dict)
+
+            with self.assertRaises((UnsupportedAlgorithmError, FormatError)):
```
Is this also different for different keys? 🤦

MVrachev replied:
Yes, for `rsa` we have a check in `securesystemslib.keys.create_signature()` which we don't have in for `ed25519` and `ecdsa`:
![image](https://user-images.githubusercontent.com/16246778/108843900-fe55e480-75e3-11eb-92a9-ef864d581da8.png)

but when we call `ed25519_keys.create_signature` or `ecdsa_keys.create_signature` there is a check inside those
functions for the `scheme` argument and that's how a FormatError is received. 

Do you think I should fix those problems here in this or?
Or in a different pr or issue?
Or even should I fix it now given that I am investigating how to replace the schemas?

lukpueh replied:
Thanks for the detailed report! Let's just add this to the wishlist in #191.

lukpueh commented on tests/test_signer.py:
```diff
@@ -0,0 +1,88 @@
+
+            # Supplying an incorrect number of arguments.
+            with self.assertRaises(TypeError):
+                sslib_signer.sign()
```
What's the reason for this test on this particular method? It seems like we could do this with any function but we don't. I think we don't need it here either.

MVrachev replied:
Okay, will remove it.

lukpueh commented on tests/test_signer.py:
```diff
@@ -0,0 +1,88 @@
+            # Verify signature
+            verified = KEYS.verify_signature(scheme_dict, sig_obj.to_dict(),
+                    self.DATA)
+            self.assertTrue(verified, "Incorrect signature.")
```
Minor nit: This module has a mix of single and double quotes. Please choose one consistently, [preferably double quotes](https://github.com/secure-systems-lab/code-style-guidelines/blob/master/python.md#quotes).

lukpueh commented on tests/test_signer.py:
```diff
@@ -0,0 +1,88 @@
+        }
+        sig_obj = Signature.from_dict(signature_dict)
+
+        self.assertEqual(signature_dict, sig_obj.to_dict())
```
Minor nit: It's better to use `assertDictEqual`, because it has a more informative error message.

lukpueh reviewed (approved):
Thanks for fixing up the docstrings, @MVrachev! I have one last set of comments in the new test module. Please fix and we are good to merge. :)

MVrachev commented:
Updated the pr:
- replaced ' with " in both `signer.py` and `test_signer.py`.
- fixed some comments 
- addressed all other Lukas comments.

I will need time to get used to the new code style guidelines.
Sorry for creating so much work in reviewing for you @lukpueh.

lukpueh commented:
> I will need time to get used to the new code style guidelines.
Sorry for creating so much work in reviewing for you @lukpueh.

No worries, I need to look things up in the new guidelines too. ;)
